### PR TITLE
kubeadm: Add a --ca-cert-path flag to kubeadm join

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

This PR makes it possible to customize where the CA file is written

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```
@pires @mikedanese @dmmcquay @jbeda @errordeveloper

### CODE REVIEW

**Comment by mikedanese** [comment]
What issue is this related to? Have people requested customizing this?

**Comment by luxas** [comment]
> Have people requested customizing this?

@mikedanese No, because no one is using this yet. But I think it makes sense to drop the envparam dependency of `kubeadm join` and have this CA cert path configurable.
Is it possible for you to approve this? I have the full proposal for kubeadm here: https://github.com/kubernetes/kubeadm/pull/156

**Comment by luxas** [comment]
@pires @mikedanese Updated

File: cmd/kubeadm/app/cmd/join.go [hunk 1]
```diff
@@ -78,6 +78,11 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 		"skip preflight checks normally run before modifying the system",
 	)
 
+	cmd.PersistentFlags().StringVar(
+		&cfg.CACertPath, "ca-cert-path", kubeadmapiext.DefaultCACertPath,
+		"path where the ca certificate for the master should be stored",
```
Comment (pires) [inline cmd/kubeadm/app/cmd/join.go]: CA certificate for the cluster not the master.

**Review by pires — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
@k8s-bot unit test this

**Review by pires — APPROVED** [review]
The code looks good but I agree we're adding functionality that originated in a WIP document (kubeadm phases design proposal) and not from actual need.

File: hack/verify-flags/exceptions.txt [hunk 2]
```diff
@@ -40,12 +40,14 @@ cluster/gce/trusty/configure-helper.sh:    sed -i -e "s@{{ *pillar\.get('storage
 cluster/gce/trusty/configure-helper.sh:  sed -i -e "s@{{pillar\['allow_privileged'\]}}@true@g" "${src_file}"
 cluster/gce/util.sh:    local node_ip=$(gcloud compute instances describe --project "${PROJECT}" --zone "${ZONE}" \
 cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:            context['pillar'] = {'num_nodes': get_node_count()}
+cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:    ca_cert_path = layer_options.get('ca_certificate_path')
```
Comment (mikedanese) [inline hack/verify-flags/exceptions.txt]: ??? where did these come from?

Comment (luxas) [inline hack/verify-flags/exceptions.txt]: No idea. Autogenerated code, verify failed on this :)

**Comment by luxas** [comment]
@k8s-bot verify test this

File: hack/verify-flags/exceptions.txt (hunk 2 above)
Comment (ixdy) [inline hack/verify-flags/exceptions.txt]: oh, probably adding the `ca-cert-path` flag made the script think that these old usages were incorrect.

I kinda hate `verify-flags`. its behavior is quite mysterious.

**Comment by luxas** [comment]
@k8s-bot verify test this
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this

File: hack/verify-flags/exceptions.txt (hunk 2 above)
Comment (luxas) [inline hack/verify-flags/exceptions.txt]: Yeah, maybe we should just remove it... don't know even what it's used for...

**Comment by luxas** [comment]
@ixdy Do you know why the verify test constantly fails like this:
```
I0221 13:33:40.386]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/client/listers/policy/v1beta1
I0221 13:33:40.899]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/client/listers/rbac/v1alpha1
I0221 13:33:40.899]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/client/listers/rbac/v1beta1
I0221 13:33:40.900]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/client/listers/storage/v1beta1
I0221 13:33:40.901]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/util
I0221 13:33:40.902]   godep: dirty working tree (please commit changes): /tmp/verify-staging-client-go.YCzei/src/k8s.io/kubernetes/pkg/util/parsers
I0221 13:33:40.904]   godep: error loading dependencies
I0221 13:33:45.406] FAILED   hack/make-rules/../../hack/verify-staging-client-go.sh	221s
```

cc @deads2k @sttts @kubernetes/sig-api-machinery-misc

**Comment by ixdy** [comment]
no idea. cc @caesarxuchao

**Comment by luxas** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot verify test this
@k8s-bot unit test this
@k8s-bot non-cri e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this

**Comment by caesarxuchao** [comment]
Can you add a `git diff` at the beginning of hack/verify-staging-client-go.sh to see what files are changed? I guess another verification script didn't clean up the changes it made.

cc @sttts

**Comment by sttts** [comment]
@caesarxuchao will take a look tomorrow.

**Comment by luxas** [comment]
@k8s-bot ok to test

**Comment by luxas** [comment]
rebased once again... hopefully https://github.com/kubernetes/kubernetes/issues/41986 fixes the verification flake

**Comment by ixdy** [comment]
/approve

I still hate verify-flags

**Comment by luxas** [comment]
@ixdy Yeah. What's wrong with Jenkins BTW?

**Comment by ixdy** [comment]
@luxas too many builds that take too long and too many open PRs.

**Comment by luxas** [comment]
@ixdy I see

**Comment by sttts** [comment]
I believe to have found why the godep call fails: we have the following unstaged change in the tree. Not sure we have any error reporting for that in CI. We should. But of course hack/verify-update-client-go.sh should not fall over it.
```
+ git status
I0223 16:54:14.760]   On branch test
I0223 16:54:14.760]   Changes not staged for commit:
I0223 16:54:14.760]     (use "git add <file>..." to update what will be committed)
I0223 16:54:14.760]     (use "git checkout -- <file>..." to discard changes in working directory)
I0223 16:54:14.760]   
I0223 16:54:14.761]   	modified:   ../../../../cmd/kubeadm/app/apis/kubeadm/v1alpha1/zz_generated.defaults.go
I0223 16:54:14.761]   
I0223 16:54:14.761]   Untracked files:
I0223 16:54:14.761]     (use "git add <file>..." to include in what will be committed)
```

**Comment by luxas** [comment]
@sttts I'm so grateful for that finding! Thank you!
Now this should be fixed... :)

**Comment by luxas** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by luxas** [comment]
@k8s-bot kubemark e2e test this

**Pull request merged**